### PR TITLE
#16296 - Self-hosting - Initial Draft

### PR DESCRIPTION
**Pull request by jnikhila**
The initial draft is based on the outline approved and available [here](https://github.com/appsmithorg/appsmith-docs/blob/v1.4/.github/drafts/16296/16296_self-hosting_outline.md). Some pointers:
* The quotes will be added to GitBook as hints
* The links are created as placeholders, and the contents will be replaced on GitBook with the appropriate cross-reference of page/section

### CODE REVIEW

File: .github/drafts/16296/16296_self-hosting.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Self-hosting
+Self-hosting is hosting and managing service directly by the owner on your own server rather than relying on a third-party service provider. Some examples include web hosting, file hosting, and email hosting. Self-hosting is often seen as a more secure and private alternative to using third-party services, as the user has complete control over their data and how it's used.
```
Comment (Pranay105) [inline .github/drafts/16296/16296_self-hosting.md:2]: I don't think we need to explain self-hosting as a general concept. We should talk about Self-hosting in the context of Appsmith from the start. One line for self-hosting is enough.

File: .github/drafts/16296/16296_self-hosting.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+Appsmith is a low-code, open source app development tool that offers a fast and easy way to develop apps. It provides a [cloud-hosted version](https://app.appsmith.com)[cloud-hosted] and the option to set up your own instance on some popular [platforms](# Deploy).
+
+## Cloud-hosted
+Cloud-hosted version is available over the internet, which makes it easy for users to sign up and start building apps. It's a perfect solution for users who don't have their own infrastructure setup. You can sign up on appsmith.com and follow the [Quick start guide](start-building) to get started.
+
+## Self-hosted
+You can choose to deploy Appsmith on your own infrastructure. It's often useful when you would want to plug it into existing infrastructure without the need to maintain a different setup. It enables you to take advantage of what your infrastructure has to offer and extend it to Appsmith.
```
Comment (Pranay105) [inline .github/drafts/16296/16296_self-hosting.md:10]: You can add one line for explaining self-hosting here.

File: .github/drafts/16296/16296_self-hosting.md (hunk 1 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:2]: We need to explain the self-hosting concept to tie it up with the whole story. Also, as per our doc structure the introduction will explain the concept that we are focusing on. the paragraph explains what is self-hosting, some examples and why users prefer self-hosting. These details are further elaborated in self-hosted section for Appsmith.

File: .github/drafts/16296/16296_self-hosting.md (hunk 2 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:10]: This section talks about self-hosting for Appsmith. And expands on the thought that was added to the introduction. The self-hosting as a general concept should be at the start of the documentation. Let's discuss this if needed.

File: .github/drafts/16296/16296_self-hosting.md (hunk 1 above)
Comment (danciaclara) [reply .github/drafts/16296/16296_self-hosting.md:2]: Nikhila, sure, we need an Introduction, but I agree with Pranay that we do not need to explain what is self-hosting at this level.  The introduction can be around letting the users know that they can also self-host Appsmith in addition to our Cloud service if they want to take care of upgrades, scaling, security etc on their own. You can directly go to the second para mentioning that there are two ways users can use Appsmith -Cloud and Self hosted (which is the Introduction). In any case, in the second para, you are talking about deploying Appsmith on your own infra. IMO, this page is about self-hosting the Appsmith platform so the introduction can be around the two ways to use Appsmith followed by the benefits of self hosting  Appsmith.

Comment (danciaclara) [reply .github/drafts/16296/16296_self-hosting.md:2]: Maybe club the three lines into one as the intro line. something like

You can host and manage Appsmith on your own server if you want to have complete control of your data, upgrades, security, and other information.

File: .github/drafts/16296/16296_self-hosting.md (hunk 2 above)
Comment (danciaclara) [reply .github/drafts/16296/16296_self-hosting.md:10]: We can skip this line "Appsmith is a low-code, open source app development tool that offers a fast and easy way to develop apps."  and directly mention - there are two ways to use the Appsmith platform.

File: .github/drafts/16296/16296_self-hosting.md (hunk 1 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:2]: Please review if the below works:
`Self-hosting Appsmith means hosting and managing Appsmith directly on your server rather than relying on third-party service providers. Self-hosting is often seen as more secure and private, as you have complete control over your data and how it's used.`
@Pranay105 and @danciaclara - Please let me so that I can update the doc.

File: .github/drafts/16296/16296_self-hosting.md (hunk 2 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:10]: done.

**Review by Pranay105 — APPROVED** [review]

File: .github/drafts/16296/16296_self-hosting.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+Appsmith provides a [secured environment](Link to security page) for cloud-hosted and self-hosted instances by encrypting sensitive information like database credentials using [AES 256 Encryption](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard). For self-hosted instances, it also supports setting up [SSL certificates](Link to SSL & Custom Domain page - Setting up SSL section) via [LetsEncrypt](https://letsencrypt.org/) during installation.
+
+#### Internal database
+A self-hosted instance could allow you to integrate with an internal database and use it to build apps.
+
+> Using a cloud-hosted instance would require the ability to access the internal databases. You would be required to use tunneling services like [ngrok](Link to Ngrok doc on Appsmtih) to expose your internal database to the internet and build apps.
```
Comment (sharat87) [inline .github/drafts/16296/16296_self-hosting.md:24]: I'm not sure I understand the first sentence. Can you rephrase/elaborate please?

Also, `ngrok` is not the only way. They cal also whitelist Appsmith cloud's external IP addresses, and allow incoming connections from just those IPs. This method is also present in our docs, and doesn't require tunneling.

File: .github/drafts/16296/16296_self-hosting.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+Appsmith uses MongoDB for storing app data and Redis for storing session information. You may wish to integrate with custom MongoDB or Redis servers hosted on your infrastructure. For this, you would have to [modify the configuration of environment variables and point them to the custom MongoDB/Redis] (Link to Custom MongoDB and Redis Page) instances.
```
Comment (sharat87) [inline .github/drafts/16296/16296_self-hosting.md:26]: We've started using Redis for non-critical caches as well, very recently.

File: .github/drafts/16296/16296_self-hosting.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+#### Updates
+You can configure more flexible updates, schedules, and other maintenance activities for a self-hosted instance. You can follow the steps to configure [maintenance window and schedule auto updates](Link to Maintenance window page).
+
+ > The self-hosted version isn't air-gapped and pings a cloud service to fetch release notes about the product updates. It helps you stay updated with the latest releases. If you would like a fully air-gapped deployment, please write to support@appsmith.com.
```
Comment (sharat87) [inline .github/drafts/16296/16296_self-hosting.md:31]: Can we change `pings a cloud service` to `reaches out to Appsmith cloud`?

Comment (sharat87) [reply .github/drafts/16296/16296_self-hosting.md:31]: Also, let's add `among other things`. Release notes is not the only thing we rely on the cloud-services for.

File: .github/drafts/16296/16296_self-hosting.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+### Deploy
+You can choose from various popular platforms to deploy Appsmith, like Docker, Kubernetes, AWS, Digital Ocean, and more.
+
```
Comment (sharat87) [inline .github/drafts/16296/16296_self-hosting.md:35]: Let's bring this line here?

```
Appsmith officially supports docker-based installations
```

**Review by sharat87 — COMMENTED** [review]
I left a few minor comments/thoughts, but overall looks good.

File: .github/drafts/16296/16296_self-hosting.md (hunk 3 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:24]: Tunneling was just an example. I'll reword it.

File: .github/drafts/16296/16296_self-hosting.md (hunk 4 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:26]: I'll add that.

File: .github/drafts/16296/16296_self-hosting.md (hunk 5 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:31]: Sure.

File: .github/drafts/16296/16296_self-hosting.md (hunk 6 above)
Comment (jnikhila) [reply .github/drafts/16296/16296_self-hosting.md:35]: Okay

**Review by jnikhila — COMMENTED** [review]
Fixed review comments.

**Review by sharat87 — APPROVED** [review]
